### PR TITLE
Configurable default member server connection limit

### PR DESCRIPTION
**Pull request by hthompson6**
As an a10-neutron-lbaas user, I want to be able to create members with connection limits.

### CODE REVIEW

File: a10_neutron_lbaas/v1/handler_member.py [hunk 1]
```diff
@@ -35,14 +36,17 @@ def _create(self, c, context, member):
         server_ip = self.neutron.member_get_ip(context, member,
                                                c.device_cfg['use_float'])
         server_name = self._meta_name(member, server_ip)
-
+        conn_limit = c.device_cfg['conn-limit']
```
Comment (Cedev) [inline a10_neutron_lbaas/v1/handler_member.py]: What happens if conn-limit isn't specified?

File: a10_neutron_lbaas/v2/handler_member.py [hunk 2]
```diff
@@ -47,14 +48,17 @@ def _create(self, c, context, member):
         server_ip = self.neutron.member_get_ip(context, member,
                                                c.device_cfg['use_float'])
         server_name = self._meta_name(member, server_ip)
-
+        conn_limit = c.device_cfg['conn-limit']
```
Comment (Cedev) [inline a10_neutron_lbaas/v2/handler_member.py]: What happens if conn-limit isn't specified?

File: a10_neutron_lbaas/v1/handler_member.py [hunk 3]
```diff
@@ -35,14 +36,17 @@ def _create(self, c, context, member):
...
+        if conn_limit > 8000000 or conn_limit < 1:
+            raise a10_ex.ConnLimitOutOfBounds("Conn-limit must be >1 and <8000000")
```
Comment (Cedev) [inline a10_neutron_lbaas/v1/handler_member.py]: Is there something more sane to do than raise an exception?

File: a10_neutron_lbaas/v2/handler_member.py [hunk 4]
```diff
@@ -47,14 +48,17 @@ def _create(self, c, context, member):
...
+        if conn_limit > 8000000 or conn_limit < 1:
+            raise a10_ex.ConnLimitOutOfBounds("Conn-limit must be >1 and <8000000")
```
Comment (Cedev) [inline a10_neutron_lbaas/v2/handler_member.py]: Is there something more sane to do than raise an exception?

File: a10_neutron_lbaas/v1/handler_member.py (hunk 1 above)
Comment (hthompson6) [inline a10_neutron_lbaas/v1/handler_member.py]: There is a default value specified that covers this (I had forgotten to add it to the previous commit). Worse case, it hits the exception and fails outright.

File: a10_neutron_lbaas/v1/handler_member.py (hunk 3 above)
Comment (hthompson6) [inline a10_neutron_lbaas/v1/handler_member.py]: Possibly, as this deals with the connection limit to servers, I didn't want to fail quietly and set a default value. I believe that the user should be alerted one way or another that they have specified an out of bounds connection limit. When in invalid limit is set on the WebUI, an error is raised that halts the connection limit from being changed and informs the user right away.

File: a10_neutron_lbaas/etc/defaults.py [hunk 5]
```diff
@@ -37,6 +37,7 @@
 DEVICE_OPTIONAL_DEFAULTS = {
     "protocol": "https",
     "port": 443,
+    "conn-limit": 8000000,
```
Comment (dougwig) [inline a10_neutron_lbaas/etc/defaults.py]: There isn't a 0 or -1 value that disables this instead?

File: a10_neutron_lbaas/v1/handler_member.py [hunk 6]
```diff
@@ -35,14 +36,17 @@ def _create(self, c, context, member):
...
         status = c.client.slb.UP
         if not member['admin_state_up']:
             status = c.client.slb.DOWN
 
         try:
             server_args = {'server': self.meta(member, 'server', {})}
             c.client.slb.server.create(server_name, server_ip,
+                                       conn_limit,
```
Comment (dougwig) [inline a10_neutron_lbaas/v1/handler_member.py]: refer to acos-client review for badness here

File: a10_neutron_lbaas/v1/handler_member.py [hunk 7]
```diff
@@ -37,14 +37,15 @@ def _create(self, c, context, member):
                                                c.device_cfg['use_float'])
         server_name = self._meta_name(member, server_ip)
         conn_limit = c.device_cfg['conn-limit']
```
Comment (mdurrant-b3) [inline a10_neutron_lbaas/v1/handler_member.py]: Can we have this be "(server|member|)-conn-limit" or similar so we know what "conn-limit" refers to?

**Comment by hthompson6** [comment]
check

File: a10_neutron_lbaas/etc/defaults.py (hunk 5 above)
Comment (mdurrant-b3) [inline a10_neutron_lbaas/etc/defaults.py]: Discussed this w/ Hunter - as there's no default on the API side and no method of disabling it, leaving the documented system default is the best practice. Perhaps submit a FR to API team to correct this behavior?

File: a10_neutron_lbaas/v1/handler_member.py (hunk 1 above)
Comment (mdurrant-b3) [inline a10_neutron_lbaas/v1/handler_member.py]: @Cedev Look 2 lines above where we do the same thing.  My only concern here is if we change the default-"fill in" behavior, these break spectacularly.

**Comment by hthompson6** [comment]
check

File: a10_neutron_lbaas/etc/defaults.py (hunk 5 above)
Comment (mdurrant-b3) [inline a10_neutron_lbaas/etc/defaults.py]: We settled on "if unspecified, don't put it in the request" - this is the best of both worlds. Gonna close this request to kick the test infrastructure.

**Comment by mdurrant-b3** [comment]
As mentioned above, closing/opening to poke test infrastructure.

**Comment by hthompson6** [comment]
check

**Comment by mdurrant-b3** [comment]
image
I thought intended functionality was setting the server port connection limit as opposed to the member server itself.
If so, we need to set the right one. If not, this works as expected and can be merged. Waiting to find out before I vote up or down.

**Comment by mdurrant-b3** [comment]
:+1: 
Setting conn-limit to -1 results in AXAPI default
Setting conn-limit to 0 results in AXAPI default
Setting conn-limit to 8000000000 (8 billion, not 8 million) results in AXAPI default
Setting conn-limit to value between 1-8000000 results in that value being set on the server connection limit properties.

**Pull request merged**